packages/nitro-protocol/contracts/MultiAssetHolder.sol

```diff
@@ -243,10 +243,280 @@ contract MultiAssetHolder is IMultiAssetHolder, StatusManager {
+        //  We rely on the assumption that the indices are strictly increasing.
+        //  This allows us to iterate over the destinations in order once, continuing until we hit the first index, then the second etc.
+        //  If the indices were to decrease, we would have to start from the beginning: doing a full search for each index.
+
+        // copy allocations
+        newSourceAllocations = new Outcome.Allocation[](sourceAllocations.length);
+        newTargetAllocations = new Outcome.Allocation[](targetAllocations.length);
+        for (uint256 i = 0; i < sourceAllocations.length; i++) {
+            newSourceAllocations[i].destination = sourceAllocations[i].destination;
+            newSourceAllocations[i].amount = sourceAllocations[i].amount;
+            newSourceAllocations[i].metadata = sourceAllocations[i].metadata;
+            newSourceAllocations[i].allocationType = sourceAllocations[i].allocationType;
+        }
+        for (uint256 i = 0; i < targetAllocations.length; i++) {
+            newTargetAllocations[i].destination = targetAllocations[i].destination;
+            newTargetAllocations[i].amount = targetAllocations[i].amount;
+            newTargetAllocations[i].metadata = targetAllocations[i].metadata;
+            newTargetAllocations[i].allocationType = targetAllocations[i].allocationType;
+        }
+
+        // compute how much the source can afford for the target
+        uint256 sourceSurplus = initialHoldings;
+        for (
+            uint256 sourceAllocationIndex;
+            sourceAllocationIndex < indexOfTargetInSource;
+            sourceAllocationIndex++
+        ) {
+            if (sourceSurplus == 0) break;
+            uint256 affordsForDestination = min(
+                sourceAllocations[sourceAllocationIndex].amount,
+                sourceSurplus
+            );
+            sourceSurplus -= affordsForDestination;
+        }
+
+        uint256 targetSurplus = min(sourceSurplus, sourceAllocations[indexOfTargetInSource].amount);
+
+        bytes32[] memory guaranteeDestinations = decodeGuaranteeData(
+            sourceAllocations[indexOfTargetInSource].metadata
+        );
+
+        for (uint256 j = 0; j < guaranteeDestinations.length; j++) {
+            if (targetSurplus == 0) break;
+            for (uint256 i = 0; i < newTargetAllocations.length; i++) {
+                if (targetSurplus == 0) break;
+                // search for it in the allocation
+                if (guaranteeDestinations[j] == newTargetAllocations[i].destination) {
+                    // if we find it, compute new amount
+                    uint256 affordsForDestination = min(targetAllocations[i].amount, targetSurplus);
+                    // decrease surplus by the current amount regardless of hitting a specified index
+                    targetSurplus -= affordsForDestination;
+                    if ((indices.length == 0) || ((k < indices.length) && (indices[k] == i))) {
```

Comment:
Earmarking for #3728. Can the intent / logic of this check hide behind some nicely named internal pure function?

Reply:
good idea!

Reply:
Let's circle back to this, for sure. It's also something that should be applied to master, regardless of the fate of this branch.